[Studio Task] Retry 2: Automated feedback generated based on failure. (SESSION-01)

@google-jules **Action Required**

### 🎯 Intent
Retry 2: Automated feedback generated based on failure.

### 📂 Context
* **Task ID:** `SESSION-01`
* **Focus Files:**
- `studio/memory.py`: Context file
- `tests/test_router_diversion.py`: Context file
- `tests/test_lifecycle_manager.py`: Context file
- `tests/test_cognitive_tunneling.py`: Context file
- `tests/test_architect_veto.py`: Context file
- `tests/test_context_slicing.py`: Context file
- `tests/test_escalation.py`: Context file
- `tests/test_orchestrator.py`: Context file
- `tests/test_tdd_loop.py`: Context file
- `tests/test_max_retry.py`: Context file
- `tests/test_refactor_loop.py`: Context file

### 📜 Constraints (MUST FOLLOW)
- [ ] Analyze the evidence provided in the logs. Generate a Virtual Patch to fix these specific errors.
- [ ] CRITICAL FEEDBACK FROM PREVIOUS ATTEMPT: Functional Verification Failed.

EVIDENCE SNIPPETS:
```
test modules/packages have valid Python names.
Traceback:
/usr/local/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/test_refactor_loop.py:5: in <module>
    from studio.memory import (
studio/memory.py:16: in <module>
    from pydantic import BaseModel, Field, HttpUrl, validator
E   ModuleNotFoundError: No module named 'pydantic'
=========================== short test summary info ============================
ERROR tests/test_router_diversion.py
ERROR tests/test_lifecycle_manager.py
ERROR tests/test_architect_veto.py
ERROR tests/test_cognitive_tunneling.py
ERROR tests/test_context_slicing.py
ERROR tests/test_escalation.py
ERROR tests/test_orchestrator.py
ERROR tests/test_tdd_loop.py
ERROR tests/test_max_retry.py
ERROR tests/test_refactor_loop.py
!!!!!!!!!!!!!!!!!!! Interrupted: 10 errors during collection !!!!!!!!!!!!!!!!!!!
============================== 10 errors in 0.39s ==============================

```

INSTRUCTION: Generate a fix (Virtual Patch) specifically addressing these errors. Ensure you adhere to the TDD Green phase requirements.

### 🔍 Relevant Logs / Evidence
```text
No logs provided.

```

---

*Generated by AI Agent Studio (Orchestrator)*

--- a/studio/subgraphs/engineer.py
+++ b/studio/subgraphs/engineer.py
@@ -432,6 +432,10 @@ async def node_qa_verifier(state: AgentState) -> Dict[str, Any]:
              raise RuntimeError("Failed to setup workspace")
 
         # 3. Install Deps
+        if "requirements.txt" in patched_files:
+            # Prefer requirements.txt for full project dependencies if available
+            sandbox.run_command("pip install -r requirements.txt")
+
         sandbox.install_dependencies(["pytest", "mock"]) # Basic deps
 
         # 4. Run Tests

--- a/tests/test_qa_verifier_infra.py
+++ b/tests/test_qa_verifier_infra.py
@@ -72,3 +72,54 @@ async def test_node_qa_verifier_ensures_infra_files():
         # EXPECTATION: These should be included even if not explicitly in context or patch
         assert "pytest.ini" in setup_call_args, "pytest.ini was NOT included in sandbox"
         assert "tests/test_logic.py" in setup_call_args, "tests/test_logic.py was NOT included in sandbox"
+
+@pytest.mark.asyncio
+async def test_node_qa_verifier_installs_requirements():
+    # 1. Setup State: requirements.txt is present
+    jules_data = JulesMetadata(
+        status="VERIFYING",
+        active_context_slice=ContextSlice(files=["requirements.txt"]),
+        generated_artifacts=[
+            CodeChangeArtifact(
+                diff_content="diff --git a/requirements.txt b/requirements.txt\n--- a/requirements.txt\n+++ b/requirements.txt\n@@ -1,1 +1,2 @@\n pydantic\n+new-dep",
+                change_type="MODIFY"
+            )
+        ]
+    )
+    state: AgentState = {
+        "messages": [],
+        "jules_metadata": jules_data,
+        "system_constitution": "",
+        "next_agent": None
+    }
+
+    # 2. Mock Dependencies
+    with patch("studio.subgraphs.engineer.DockerSandbox") as MockSandbox, \
+         patch("studio.subgraphs.engineer.apply_virtual_patch") as mock_apply, \
+         patch("studio.subgraphs.engineer.extract_affected_files") as mock_extract, \
+         patch("os.path.exists") as mock_exists, \
+         patch("builtins.open") as mock_open:
+
+        mock_extract.return_value = ["requirements.txt"]
+
+        # mock_apply returns patched files including requirements.txt
+        mock_apply.return_value = {"requirements.txt": "pydantic\nnew-dep"}
+
+        mock_exists.return_value = True # Simulate all files exist
+
+        # Mock open
+        mock_file = MagicMock()
+        mock_file.__enter__.return_value.read.return_value = "pydantic"
+        mock_open.return_value = mock_file
+
+        mock_sandbox_inst = MagicMock()
+        MockSandbox.return_value = mock_sandbox_inst
+        mock_sandbox_inst.setup_workspace.return_value = True
+        mock_sandbox_inst.run_pytest.return_value = MagicMock(passed=True, error_log=None)
+        mock_sandbox_inst.run_command.return_value = MagicMock(exit_code=0)
+
+        # 3. Execute the node
+        await node_qa_verifier(state)
+
+        # 4. Verify pip install -r requirements.txt was called
+        mock_sandbox_inst.run_command.assert_any_call("pip install -r requirements.txt")

